hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java

```diff
@@ -279,24 +289,22 @@ private void populateResultFromNearCache(Collection<?> keys, List<Object> result
         }
     }
 
-    private Map<Object, Long> getNearCacheReservations(Collection<?> nearCacheKeys, Map<Object, Data> keyMap) {
-        Map<Object, Long> reservations = createHashMap(nearCacheKeys.size());
-        for (Object key : nearCacheKeys) {
-            Data keyData = serializeKeys ? (Data) key : keyMap.get(key);
-            long reservationId = tryReserveForUpdate(key, keyData);
+    private void createNearCacheReservations(Collection<?> dataKeys, Map<Object, Long> reservations) {
+        for (Object key : dataKeys) {
+            long reservationId = tryReserveForUpdate(key, (Data) key);
             if (reservationId != NOT_RESERVED) {
                 reservations.put(key, reservationId);
             }
         }
-        return reservations;
     }
 
-    private void populateResultFromRemote(int currentSize, List<Object> resultingKeyValuePairs, Map<Object, Long> reservations) {
+    private void populateResultFromRemote(int currentSize, List<Object> resultingKeyValuePairs, Map<Object, Long> reservations,
+                                          Map<Data, Object> reverseKeyMap) {
         for (int i = currentSize; i < resultingKeyValuePairs.size(); i += 2) {
             Data keyData = (Data) resultingKeyValuePairs.get(i);
             Data valueData = (Data) resultingKeyValuePairs.get(i + 1);
 
-            Object ncKey = serializeKeys ? keyData : toObject(keyData);
+            Object ncKey = serializeKeys ? keyData : reverseKeyMap.get(keyData);
```

Comment:
By omitting the `toObject()` call we:
* save an additional de-serialization
* do not clone the key object and run into the instance identity issue